### ISSUE TITLE
StopIteration: attempt to get argmax of an empty sequence

### ISSUE THREAD

**Issue by nealzGu** [fizyr/keras-retinanet#49]
Hi again,

Im getting the following error:
(most recent call last):
  File "examples/train_pascal.py", line 106, in <module>
    keras.callbacks.ReduceLROnPlateau(monitor='val_loss', factor=0.1, patience=10, verbose=1, mode='auto', epsilon=0.0001, cooldown=0, min_lr=0),
  File "/usr/local/lib/python2.7/dist-packages/keras/legacy/interfaces.py", line 87, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/keras/engine/training.py", line 2046, in fit_generator
    generator_output = next(output_generator)
  File "/usr/local/lib/python2.7/dist-packages/keras/utils/data_utils.py", line 518, in get
    raise StopIteration(e)
StopIteration: attempt to get argmax of an empty sequence

From [here](https://github.com/delftrobotics/keras-retinanet/issues/34) it seems as if its a problem of no significant overlap? I checked the image and there should be eleven bboxes in there, covering almost all the space, so Im not sure if that's really the error here.

Any ideas?

I will remove the image for now and continue training, to check if its just the image...

**Comment by nealzGu** [comment]
not an issue of a specific image.

**Comment by LaurensHagendoorn** [comment]
I also just ran into the same error:

`File "/home/baloo/Develop/Laurens/retinanet/keras_retinanet/preprocessing/anchors.py", line 30, in anchor_targets
    argmax_overlaps_inds = np.argmax(overlaps, axis=1)
  File "/home/baloo/anaconda3/envs/-/lib/python3.6/site-packages/numpy/core/fromnumeric.py", line 963, in argmax
    return _wrapfunc(a, 'argmax', axis=axis, out=out)
  File "/home/baloo/anaconda3/envs/-/lib/python3.6/site-packages/numpy/core/fromnumeric.py", line 57, in _wrapfunc
    return getattr(obj, method)(*args, **kwds)
ValueError: attempt to get argmax of an empty sequence`

**Comment by nealzGu** [comment]
this morning I trained on a really small data set (60 images), just for testing - it went through!

With whole data set (4k images) it crashed at a random step during epoch 1...

**Comment by LaurensHagendoorn** [comment]
I have some images that do not contain any annotations (but do have an annotation file) so I suspect those could be the problem, but it is sort of hard to pin down thus far.

**Comment by nealzGu** [comment]
Im also having an Annotation file for each image... actually not sure if any of them are "empty" ... let's keep in touch to solve this!

**Comment by LaurensHagendoorn** [comment]
Nope, that's not the issue, if I print the gt_boxes, it encounters a few empty ones before the crash

**Comment by nealzGu** [comment]
Now Im training on 600 images. Its crashing at epoch 1 and 599/600. Happend three times now. Still not sure why

**Comment by hgaiser** [comment]
@LaurensHagendoorn since keras buffers multiple batches in parallel, it could be that you see it print multiple empty boxes before it will actually process one of them. I think a check should be added to prevent this error when no gt box is provided.

**Comment by zihangJiang** [comment]
I also ran into the same error :raise StopIteration(e)
StopIteration: attempt to get argmax of an empty sequence 
but I checked my data,finding one of my picture had an empty boundbox.After deleting it,everything just goes well.

**Comment by LaurensHagendoorn** [comment]
Than that probably is the problem, I will get back to you once I've fixed it. Kind of a shame, the old faster-rcnn repository did not have this problem ;)

**Comment by hgaiser** [comment]
It doesn't say so explicitly, but I incorporated this fix in the refactoring of the generators in PR https://github.com/fizyr/keras-retinanet/pull/48. I'll close this when that gets merged.

**Comment by LaurensHagendoorn** [comment]
Yeah I saw, awesome! I hope that will fix this, I will report back once I've used it. (but my server is currently not functional)

**Comment by LaurensHagendoorn** [comment]
In that new branch I am getting:  "AttributeError: module 'keras_resnet' has no attribute 'custom_objects'"

**Comment by hgaiser** [comment]
> In that new branch I am getting: "AttributeError: module 'keras_resnet' has no attribute 'custom_objects'"

Thats correct, you need [this](https://github.com/fizyr-forks/keras-resnet/tree/disable-bias) branch of `keras-resnet` for that PR. I'm not yet seeing the same results as what I got before though, so tread with caution ;)

**Comment by LaurensHagendoorn** [comment]
`Traceback (most recent call last):
  File "train_ding.py", line 61, in <module>
    model = create_model(weights=C.base_net_weights)
  File "train_ding.py", line 39, in create_model
    return ResNet50RetinaNet(image, num_classes=len(C.classes), weights=weights)
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/models/resnet.py", line 44, in ResNet50RetinaNet
    model = keras_retinanet.models.retinanet.retinanet_bbox(inputs=inputs, backbone=resnet, *args, **kwargs)
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/models/retinanet.py", line 204, in retinanet_bbox
    model = retinanet(inputs=inputs, num_classes=num_classes, *args, **kwargs)
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/models/retinanet.py", line 187, in retinanet
    submodels = default_submodels(num_classes, anchor_parameters)
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/models/retinanet.py", line 151, in default_submodels
    ('regression', default_regression_model(anchor_parameters.num_anchors())),
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/models/retinanet.py", line 99, in default_regression_model
    outputs = keras.layers.Reshape((-1, 4), name='pyramid_regression_reshape')(outputs)
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/engine/topology.py", line 602, in __call__
    output = self.call(inputs, **kwargs)
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/layers/core.py", line 391, in call
    target_shape = self.compute_output_shape(input_shape)[1:]
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/layers/core.py", line 376, in compute_output_shape
    input_shape[1:], self.target_shape)
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/layers/core.py", line 364, in _fix_unknown_dimension
    original = np.prod(input_shape, dtype=int)
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/numpy/core/fromnumeric.py", line 2498, in prod
    out=out, **kwargs)
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/numpy/core/_methods.py", line 35, in _prod
    return umr_prod(a, axis, dtype, out, keepdims)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'`

I am still getting this error now, but that's a problem for tomorrow

**Comment by hgaiser** [comment]
Do you use keras 2.0.9?

**Comment by LaurensHagendoorn** [comment]
ah! No, not anymore

**Comment by LaurensHagendoorn** [comment]
I suppose I am inpatient and should just wait untill the pull request is merged ;P
I cannot download from your sharepoint?

I also disabled the loading of the weights (In which case I assume it randomly initialises them?) which leads to this error:
`Epoch 1/200
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/utils/data_utils.py", line 579, in data_generator_task
    generator_output = next(self._generator)
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/preprocessing/generator.py", line 117, in __next__
    return self.next()
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/preprocessing/generator.py", line 133, in next
    image_group, annotations_group = self.preprocess_group(image_group, annotations_group)
  File "/home/laurenshagendoorn/deployed/stedin/retinanet/keras_retinanet/preprocessing/generator.py", line 87, in preprocess_group
    image, annotations = random_transform(image, annotations, self.image_data_generator)
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/utils/image.py", line 64, in random_transform
    cv2.rectangle(mask, (b[0], b[1]), (b[2], b[3]), (255,) * image.shape[-1], -1)
TypeError: Layout of the output array img is incompatible with cv::Mat (step[ndims-1] != elemsize or step[1] != elemsize*nchannels)

Traceback (most recent call last):
  File "train_stedin.py", line 108, in <module>
    keras.callbacks.ReduceLROnPlateau(monitor='val_loss', factor=0.1, patience=10, verbose=1, mode='auto', epsilon=0.0001, cooldown=0, min_lr=0),
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/legacy/interfaces.py", line 87, in wrapper
    return func(*args, **kwargs)
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/engine/training.py", line 2046, in fit_generator
    generator_output = next(output_generator)`

I am currently testing on a dataset that does not have any empty annotations btw.

**Comment by hgaiser** [comment]
I updated the link, something apparently went wrong in sharing.

Yeah it's probably better to be patient, but I also encourage exploring the unknown ;)

That error is strange, never seen that before. It almost looks like your image is incorrect in some way?

**Comment by LaurensHagendoorn** [comment]
the old faster-rcnn implementation could train on this set, and previously this retinanet also worked on it.
So there shouldn't really be something wrong with the image, I think it was introduced when I switched to the new branch, but I'll look into it.

**Comment by LaurensHagendoorn** [comment]
Found a new one! But this one I really am saving for tomorrow.

It appears that some of the boundingboxless images are still getting through to the augmentation. Strange, as on a previous dataset with missing boundingboxes it did get to the end of the set without errors.

```
Epoch 1/200
  12/2525 [..............................] - ETA: 23:49 - loss: 7.6793 - regression_loss: 6.5561 - classification_loss: 1.1232Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/utils/data_utils.py", line 579, in data_generator_task
    generator_output = next(self._generator)
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/preprocessing/generator.py", line 117, in __next__
    return self.next()
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/preprocessing/generator.py", line 133, in next
    image_group, annotations_group = self.preprocess_group(image_group, annotations_group)
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/preprocessing/generator.py", line 87, in preprocess_group
    image, annotations = random_transform(image, annotations, self.image_data_generator)
  File "/home/laurenshagendoorn/deployed/ding/retinanet/keras_retinanet/utils/image.py", line 77, in random_transform
    boxes[index, 0] = float(min(j))
ValueError: min() arg is an empty sequence

Traceback (most recent call last):
  File "train_ding.py", line 108, in <module>
    keras.callbacks.ReduceLROnPlateau(monitor='val_loss', factor=0.1, patience=10, verbose=1, mode='auto', epsilon=0.0001, cooldown=0, min_lr=0),
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/legacy/interfaces.py", line 87, in wrapper
    return func(*args, **kwargs)
  File "/home/laurenshagendoorn/anaconda3/lib/python3.6/site-packages/keras/engine/training.py", line 2046, in fit_generator
    generator_output = next(output_generator)
StopIteration
```

**Comment by hgaiser** [comment]
Sounds like it has a box which is empty or something. Worth looking into that annotation at least.

**Comment by LaurensHagendoorn** [comment]
Hm, maybe, this is the last file prints before crashing, seems fine. I also checked the annotation files of the other loaded files, and those also all look fine. I thought perhaps the truncation flag could be an issue (as it is the odd one out) but on my next run it crashed on one without the truncation flag.


```
<annotation>
        <folder>JPEGImages</folder>
        <filename>107678637_0_13.jpg</filename>
        <path>\\validpath\107678637_0_13.jpg</path>
        <source>
                <database>Unknown</database>
        </source>
        <size>
                <width>1200</width>
                <height>900</height>
                <depth>3</depth>
        </size>
        <segmented>0</segmented>
        <object>
                <name>thing1</name>
                <pose>Unspecified</pose>
                <truncated>1</truncated>
                <difficult>0</difficult>
                <bndbox>
                        <xmin>1</xmin>
                        <ymin>464</ymin>
                        <xmax>41</xmax>
                        <ymax>761</ymax>
                </bndbox>
        </object>
        <object>
                <name>thing1</name>
                <pose>Unspecified</pose>
                <truncated>0</truncated>
                <difficult>0</difficult>
                <bndbox>
                        <xmin>656</xmin>
                        <ymin>491</ymin>
                        <xmax>679</xmax>
                        <ymax>678</ymax>
                </bndbox>
        </object>
        <object>
                <name>thing1</name>
                <pose>Unspecified</pose>
                <truncated>0</truncated>
                <difficult>0</difficult>
                <bndbox>
                        <xmin>150</xmin>
                        <ymin>283</ymin>
                        <xmax>564</xmax>
                        <ymax>437</ymax>
                </bndbox>
        </object>
</annotation>
```

**Comment by nealzGu** [comment]
I also adopted the changes of the PR, but still got an error for empty images. I manually sorted out the empty ones and noch training works fine :)

Still, if someone figures out how to deal with this, please let us know!

**Comment by LaurensHagendoorn** [comment]
After removing the empty annotations I still get the error on the min(j), but I suppose it must be due to an incorrect annotation somewhere still.

**Comment by hgaiser** [comment]
This issue is addressed in PR https://github.com/fizyr/keras-retinanet/pull/48.

@LaurensHagendoorn if there is still an issue with data augmentation, please make a new issue.
